teeracle/src/lib.rs

```diff
@@ -93,52 +109,86 @@ pub mod pallet {
 	#[pallet::call]
 	impl<T: Config> Pallet<T> {
 		#[pallet::weight(<T as Config>::WeightInfo::add_to_whitelist())]
-		pub fn add_to_whitelist(origin: OriginFor<T>, mrenclave: [u8; 32]) -> DispatchResult {
+		pub fn add_to_whitelist(
+			origin: OriginFor<T>,
+			data_source: MarketDataSourceString,
+			mrenclave: [u8; 32],
+		) -> DispatchResult {
 			ensure_root(origin)?;
-			ensure!(!Self::is_whitelisted(mrenclave), <Error<T>>::ReleaseAlreadyWhitelisted);
-			<Whitelist<T>>::try_append(mrenclave)
-				.map_err(|_| Error::<T>::ReleaseWhitelistOverflow)?;
-			Self::deposit_event(Event::AddedToWhitelist(mrenclave));
+			ensure!(
+				!Self::is_whitelisted(data_source.clone(), mrenclave),
+				<Error<T>>::ReleaseAlreadyWhitelisted
+			);
+			<Whitelists<T>>::try_mutate(data_source.clone(), |mrenclave_vec| {
+				mrenclave_vec.try_push(mrenclave)
+			})
+			.map_err(|_| Error::<T>::ReleaseWhitelistOverflow)?;
+			Self::deposit_event(Event::AddedToWhitelist(data_source, mrenclave));
 			Ok(())
 		}
 		#[pallet::weight(<T as Config>::WeightInfo::remove_from_whitelist())]
-		pub fn remove_from_whitelist(origin: OriginFor<T>, mrenclave: [u8; 32]) -> DispatchResult {
+		pub fn remove_from_whitelist(
+			origin: OriginFor<T>,
+			data_source: MarketDataSourceString,
+			mrenclave: [u8; 32],
+		) -> DispatchResult {
 			ensure_root(origin)?;
-			ensure!(Self::is_whitelisted(mrenclave), <Error<T>>::ReleaseNotWhitelisted);
-			Whitelist::<T>::mutate(|mrenclaves| mrenclaves.retain(|m| *m != mrenclave));
-			Self::deposit_event(Event::RemovedFromWhitelist(mrenclave));
+			ensure!(
+				Self::is_whitelisted(data_source.clone(), mrenclave),
+				<Error<T>>::ReleaseNotWhitelisted
+			);
+			<Whitelists<T>>::mutate(data_source.clone(), |mrenclave_vec| {
+				mrenclave_vec.retain(|m| *m != mrenclave)
+			});
+			Self::deposit_event(Event::RemovedFromWhitelist(data_source, mrenclave));
 			Ok(())
 		}
 
 		#[pallet::weight(<T as Config>::WeightInfo::update_exchange_rate())]
 		pub fn update_exchange_rate(
 			origin: OriginFor<T>,
-			currency: CurrencyString,
+			data_source: MarketDataSourceString,
+			trading_pair: TradinPairString,
 			new_value: Option<ExchangeRate>,
 		) -> DispatchResultWithPostInfo {
 			let sender = ensure_signed(origin)?;
 			<pallet_teerex::Module<T>>::is_registered_enclave(&sender)?;
```

Comment:
here you could ensure vec argument size limits (or benchmark the size as stated above)